org-roam.el

```diff
@@ -238,39 +238,48 @@ in temp buffers.  In cases where this occurs, we do know the file path, and pass
-                                (t nil))))
-          (when link-type
-            (goto-char start)
-            (let* ((element (org-element-at-point))
-                   (begin (or (org-element-property :content-begin element)
-                              (org-element-property :begin element)))
-                   (content (or (org-element-property :raw-value element)
-                                (buffer-substring-no-properties
-                                 begin
-                                 (or (org-element-property :content-end element)
-                                     (org-element-property :end element)))))
-                   (content (string-trim content))
-                   ;; Expand all relative links to absolute links
-                   (content (org-roam--expand-links content file-path)))
-              (vector file-path
-                      (cond ((string= link-type "roam")
-                             (file-truename (expand-file-name path (file-name-directory file-path))))
-                            ((string= link-type "cite")
-                             path))
-                      link-type
-                      (list :content content :point begin)))))))))
+    (->>
+     (org-element-map (org-element-parse-buffer) 'link
+       (lambda (link)
+         (let* ((type (org-element-property :type link))
+                (path (org-element-property :path link))
+                (start (org-element-property :begin link))
+                (link-type (cond ((and (string= type "file")
+                                       (org-roam--org-file-p path))
+                                  "roam")
+                                 ((and
+                                   (require 'org-ref nil t)
+                                   (-contains? org-ref-cite-types type))
+                                  "cite")
+                                 (t nil))))
+           (when link-type
+             (goto-char start)
+             (let* ((element (org-element-at-point))
+                    (begin (or (org-element-property :content-begin element)
+                               (org-element-property :begin element)))
+                    (content (or (org-element-property :raw-value element)
+                                 (buffer-substring-no-properties
+                                  begin
+                                  (or (org-element-property :content-end element)
+                                      (org-element-property :end element)))))
+                    (content (string-trim content))
+                    ;; Expand all relative links to absolute links
+                    (content (org-roam--expand-links content file-path)))
+               (vector file-path
+                       (cond ((string= link-type "roam")
+                              (file-truename (expand-file-name path (file-name-directory file-path))))
+                             ((string= link-type "cite")
```

Comment:
I'm wondering if we couldn't distribute the `cite` at this point in the code.  That would prevent us from running an extra cond-check on every entries to check if they're a `cite` link.

Reply:
I did think about that, but I'm not sure how that would work. `org-element-map` acts like a `mapcar`, so one element can't be transformed into multiple elements (as far as I'm aware)

Reply:
I do agree though it would be ideal to do it there

Reply:
I guess we could have `org-element-map` map to a list of lists and then flatten that?

Reply:
> I did think about that, but I'm not sure how that would work. `org-element-map` acts like a `mapcar`, so one element can't be transformed into multiple elements (as far as I'm aware)

I think we might be able to get away with it with some Elisp Magic™✨, i.e., probably `pcase` and `setf`.  I'm working on it.

Reply:
Fixed in 734da3d250a965b5e0f6f33a0c6699c6aa67c19b.  Found a better way to do it.

Reply:
Nice. Looks good